### ISSUE TITLE
What's causing poor 3d reconstruction of these images?

### ISSUE THREAD

**Issue by Goldname100** [alicevision/meshroom#536]
I'm trying to reconstruct a shoe. I have taken several hundred photographs and retried multiple times. Every time I try to reconstruct the shoe, there are parts not reconstructed, some reconstructed parts are very blurry, and there's a high amount of noise. I have a screenshot of my images and the 3d reconstruction.

Is this reconstruction "normal" in the sense that this is the best I can do? Or is there something wrong? Furthermore, a big chunk of images were not used, labelled "Camera: not reconstructed"
Thank you

![image](https://user-images.githubusercontent.com/22924301/60557303-e720ff80-9cf9-11e9-85d8-dae0bbe681aa.png)

**Comment by hargrovecompany** [comment]
zoom in to the point that the "foreground" (shoe) takes up a lot more of your screen...maybe 50%.  Also, if possible, put some other objects on the table top to provide additional features.  You have features sort of floating above the shoe which means images didn't align.  I don't know in this circumstance if the white background is hurting you but I think it probably is.

**Comment by Goldname100** [comment]
I have zoomed in. See here:
![image](https://user-images.githubusercontent.com/22924301/60566860-12691600-9d1d-11e9-8e7f-4e9ec6815362.png)

I actually tried using only a mere subset of these several hundred images, and it seemed to construct a much cleaner model. However, this model still has gaping holes due to many images not being used. Why does using more images create a noisier image? 

Also, how can I prevent the discarding of entire groups of images that are taken from one side of this shoe? The shoe rests on a rotating mini studio platform with the camera completely stabilized, so I don't understand how some parts of the image are thrown out.

**Comment by hargrovecompany** [comment]
If i could see  more clearly which images are kicked out, I might be able to help....I just can't see much  in your images on the screenshot

**Comment by hargrovecompany** [comment]
out of curiosity, are you moving the shoe and leaving the camera static?

**Comment by Goldname100** [comment]
Yes, the shoe is sitting on a rotating platform. The camera is static. 

I have a screenshot of some of the pictures that were kicked out. Is this useful? 
![image](https://user-images.githubusercontent.com/22924301/60567732-c53a7380-9d1f-11e9-8a24-c85578e890f3.png)

**Comment by hargrovecompany** [comment]
ok....again, i can't see enough detail to offer much, so let me ask a couple more question.  how well distributed is the lighting?  if you have only a few light sources, the shadows (for example under the shoe laces) will be in different locations relative to the show as the table rotates.     also, did you pick the shoe up and turn it over to get the bottom, or is it a glass table?

**Comment by hargrovecompany** [comment]
assuming you have good resolution, you should be able to get a great rendering with maybe 70 images.

**Comment by hargrovecompany** [comment]
there is a thread somewhere here where a guy did a banana in full circumference, xyz.  the banana was on a glass table, and the result was really good

**Comment by Goldname100** [comment]
The first screenshot has a zoomed in picture of one specific image that was not used in reconstruction. Do you want to see another zoomed in image?

The lighting should be very well distributed. I am using a photo studio like the one shown here: 
![image](https://user-images.githubusercontent.com/22924301/60568140-ce781000-9d20-11e9-9265-4ef7976c9326.png)

Yes, I turned the shoe over, but with the good lighting, there shouldn't be any shadows I think.

**Comment by hargrovecompany** [comment]
ok....is it all all possible that the shape of the shoe changed slightly when you flipped it?

**Comment by Goldname100** [comment]
I tried reconstruction with a banana, but for some reason I kept getting Depthmap errors from this thread: https://github.com/alicevision/meshroom/issues/409#issue-416296254 (after I was done with the shoe, I wanted to look into this)

I didn't think the minor shoe deformation would matter very much? If it does I won't flip it

**Comment by hargrovecompany** [comment]
i don't know how much forgiveness is built into the software but i would guess that moving the shoe is problematic

**Comment by Goldname100** [comment]
Regardless, I did try again without flipping the shoe over at all, and the reconstruction was still very poor. 

Is it a known bug that using too many images will result in poor reconstruction? If not, do you have any other idea?

**Comment by hargrovecompany** [comment]
what are your settings?   i.e. Sift/Sift Float/Sift Upright/Akaze....    Normal/High/Ultra?

**Comment by hargrovecompany** [comment]
if you selected about 40 images taken before you flipped the shoe over (again, assuming it wasn't moved while those images were taken, i'm guessing you would get a pretty good rendering.  can you get a shot of the cameras that should have been adjacent, the last that was included and the first that was excluded?

**Comment by Goldname100** [comment]
What do you mean by "get a shot of the cameras that should have been adjacent?"

Where do I find those settings at?

**Comment by hargrovecompany** [comment]
ok....where you see the images imported on the left, you'll see the red error on the top right corner of the images not used.  i can't remember but I think it gives a green indicator on the top right if they were used.  the top left corner of the image has green if it imported, read exif info, the camera type is in the database and it's all good to try to use for the rendering.  you can also look at the 3d image and see where cameras are missing.  

settings....if you go to the bottom and click on the second module in the workflow, you'll see options for different methods of feature extraction and resolution/granularity.  Usually, out of the box settings are really good for almost everything (sift, and normal).  you can experiment with those....but the same boxes need to be checked for sift/sift float/sift upright/akaze/etc. where they appear (i think they're an option in 3 of the first 5 modfules

**Comment by hargrovecompany** [comment]
one thing that might be problematic.....the shots of the shoe that shoe the liner around the opening....that material might be catching light differently as the table rotates.  and, its somewhat featureless.

**Comment by Goldname100** [comment]
Ok interesting.

The settings have nothing clicked except for: sift. Are you saying all of those boxes should be clicked?

I think this screenshot is what you wanted to see?
![image](https://user-images.githubusercontent.com/22924301/60569734-e356a280-9d24-11e9-8ea4-fc932a57228d.png)

I am still confused why some images are included and some are excluded.

**Comment by hargrovecompany** [comment]
no....and i need to be careful because i am not one of the smart guys who built this software....sift is an algorythm, just as sift float is.  and sift upright.  and sift akaze.  the algorythms are different approaches for identifying and matching the features.  one might get a better result than another in a certain scenario.  you can include more than one, but that's about as far as i can take you.   

the other selection....i forget exactly what it's called, more or less determines how small you want to go when you identify features.  it seems to me that if an image set is problematic (flare, reflections, etc) i get a better result with lower resolution.

**Comment by natowi** [comment]
Have a look at [Reconstruction-parameters](https://github.com/alicevision/meshroom/wiki/Reconstruction-parameters). I would try with adding AKAZE in the node settings

**Comment by hargrovecompany** [comment]
natowi, you won't hurt my feelings at all if you correct anything that i said that's not accurate....

**Comment by hargrovecompany** [comment]
also, your opinion....when he flips the shoe over, even if it's pretty minor, is a change in the "shape" of the  shoe going to change the feature locations enough to blow up the image matching?

**Comment by natowi** [comment]
@hargrovecompany your advices are good. The more changes to the captured object you avoid, the better.

Maybe a setup like [this](https://forum.sketchfab.com/t/best-method-for-capturing-all-sides-of-an-object/1138/28) would be worth a try.

**Comment by hargrovecompany** [comment]
goldname100.....another quick question....have you changed zoom/focal length at all between shots?  if so, that will definitely cause a problem

**Comment by natowi** [comment]
Different focal length are supported, but should not be mixed. To use multiple settings import the images in multiple groups.

**Comment by hargrovecompany** [comment]
each specific focal length as a group?

**Comment by hargrovecompany** [comment]
this is good info.  my rig has a bunch of dslr with 18 to 55mm lens.  i was wondering how i could zoom a few of them without blowing up the  rendering

**Comment by natowi** [comment]
@hargrovecompany required the focal length is stored in the EXIF. I´ll add details to the doc.

**Comment by hargrovecompany** [comment]
natowi (apologies for hijacking).....if i have two different types of cameras, do I need to import them segregated into groups?

**Comment by natowi** [comment]
@hargrovecompany you need to make sure your images from different devices do not share the same camera SerialNumber metadata.

**Comment by hargrovecompany** [comment]
yeah, that's a problem because i have 60 raspberry pi cameras and they have no serial number in the metadata.  i havent seen any rendering issues that i think would be related.  however, i have had "ghosting" where the rendered image appears twice, when my dslr cameras were at different focal lengths from one another....it seems that it happened only with sift and akaze checked on high.   i can't submit test images if you want them

**Comment by natowi** [comment]
@hargrovecompany you can open a seperate issue on this

**Comment by Goldname100** [comment]
@hargrovecompany @natowi  Yes I have definitely changed both the focus and the zoom while taking different sets of image, but only by very little. Thank you for pointing out this mistake.

Is each zoom change supposed to be in a different group or is each focus change supposed to be in a different group?

After I import the settings in multiple groups, are there any other settings I need to change?

**Comment by natowi** [comment]
@Goldname100 I have to test this out for myself, too. When importing your images, check the image parameters in the CameraInit node and make sure the correct metadata is used.

**Comment by natowi** [comment]
Using Augment Reconstruction will create Image Groups. For each image group a new CameraInit node is being created. Adding images with different focal lengths or from differnt camera models to Image Groups will help to prevent accidental mix-ups.

**Comment by Goldname100** [comment]
@natowi I was just following the guide you gave here: https://github.com/alicevision/meshroom/wiki/Reconstruction-parameters

Following the sparse reconstruction tutorial I couldn't find a "Load Model" option at the end. How do I see the model?

**Comment by natowi** [comment]
@Goldname100 there is a minor "Load Model" bug. You might need to remove all views from the 3D viewer TOC and re-add the Meshing node (and SFM) output with a double click on the node.

**Comment by Goldname100** [comment]
Ah thanks, I got it. So, the below screenshot was done by simply using a small subset of the large images that I took. However, I am still concerned about the white junk all over the shoe. Do you know how to get rid of this?

![subset_different_parameters](https://user-images.githubusercontent.com/22924301/60615715-a4f0d000-9d84-11e9-8540-587ca977a70b.JPG)

**Comment by natowi** [comment]
In this case I would put your shoe on three spikes, so I could take pictures from below, while keeping it on my rotating table. Instead of the white underground I would choose a well structured surface.
If your spikes are thin and reflective (nails) your model should float in the air. (Here we can make use of the photogrammetry problem with reflective surfaces ;) ). This does not work all the time, but these spikes are easy to remove in Meshlab, too.

**Comment by Goldname100** [comment]
Thanks for the recommendation. I don't have much time to get some mounts for this, but it's not a big problem because I'm actually not too concerned about modelling the bottom of the shoe.

Instead, the white junk around the shoe is a greater concern to me. Will this problem be solved by not using a plain white background and simply not rotating the shoe but rather the camera itself? I'm giving it a try but the model being reconstructed is worse than before. 

Is indoor lighting allowed?

**Comment by natowi** [comment]
In this case, use a well structured ground plane for your rotation table and remove it from the final model. this will prevent those white blobs.

**Comment by Goldname100** [comment]
Ok thanks. I just found that deleting StructureFromMotion and leaving only Texture in the 3d viewpoint shows a much clearer item. Can you explain why?

**Comment by hargrovecompany** [comment]
Yes.
Structure from motion uses marker in the visual representation that you see that can be scaled to larger or smaller with the slider.  They are not intended to be part of the completed image.  They are simply features that are common among several images.  The spots you see simply represent coordinates for alignment of the images more or less for camera positioning.  Seeing a ton of features in the sfm model is good.  sometimes there are so many that it almost appears to be the textured image.

**Comment by hargrovecompany** [comment]
if you're just worried about the top of the shoe, put the shoe on a piece of newspaper on a table with good distributed lighting.  take 20 shots in a ring at the lowest level.  move up 10 degrees.  take 18 shots.  move up 10 degrees.  take 16 shots.  move up 10 degrees.  take 14 shots....etc.  make sure you are close enough so that the shoe consumes at least 50 percent of the frame.   I GUARANTEE that so long as your camera is not on auto white balance, and you don't change the zoom, you will get an awesome render on the first run using factory settings

**Comment by Goldname100** [comment]
@natowi Thanks this has been very helpful. 

I actually had one more question. Is there a way to import a previously built mesh and reconstruct only the texture with images not used for reconstructing the mesh (only slightly different)?

**Comment by natowi** [comment]
In the "PrepareDenseScene" node you can use the "imagesFolders" option to override input images. This enables to use one image set for reconstruction and do the Texturing with another set of images.
(useful for structured light scans in a rig)

**Comment by Goldname100** [comment]
@natowi Sorry I misspoke on the previous comment. After thinking a bit more, I realized what I'm actually looking for is to modify the existing texture, not rebuild it.

Basically, I'm wondering if AliceVision can do this: 

Render photos of a 3d Model -> for each rendered photo, get some corresponding modifications (I will do this) -> remap these modifications back to the 3D model by superimposing on top of the texture where the original photo was rendered (aka "back render" by superimposing a new image).

Possible that Alicevision can help?

**Comment by natowi** [comment]
@Goldname100 I do not understand what you are trying to achieve. 

My previously listed options are the only way to modify the texture.

**Comment by Goldname100** [comment]
@natowi Sorry for the late response. I'm still trying to find a way to modify the texture by projecting changes onto them. Do you have an API to do this perhaps?

**Comment by natowi** [comment]
@Goldname100 No, but you can use Blender for this. Use the paint tools in Blender to edit your texture and render it to a camera from your alembic file (previously reconstructed in Meshroom).
You can use [meshroom2blender](https://github.com/tibicen/meshroom2blender).

**Comment by cnlohr** [comment]
Did anyone ever figure out the actual reason behind why many images aren't used for reconstruction?  I am having a very similar issue, where I went back and with exactly the same camera settings I took more pictures of the troubled area, to no avail.  It just keeps throwing out the pictures that fill in the detail I'm after.

I've also tried combinations of SIFT, AKAZE and trying wildly different values for describer density and describer quality.  It's very frustrating to see how many cameras the algo doesn't use even when I provide a LOT of overlap.
